CONTRIBUTING.md

```diff
@@ -210,6 +214,231 @@ it is helpful to use [npm link](https://docs.npmjs.com/cli/link).
+
+### Organizing Tests
+
+It is common practice, to group automated tests in categories depending on how 
+broad their scope is, and how complex their execution requirements are. 
+This separation often follows the 
+[testing pyramid](https://martinfowler.com/bliki/TestPyramid.html). 
+In DC/OS UI we use three levels of testing: Unit, Integration, and System.
+
+[Unit Tests](#unit-testing) are intended to verify functionality at the minimal 
+level of isolation, for that purpose, they tend to be small, and mock/stub all 
+its dependencies. They also tend to execute fast and in a reliable way.
+
+Given a system is composed of a collection of components, that when tested 
+in isolation perform correctly, but often fail to do so together, we use both 
+integration testing and system testing to verify behavior at an increasing 
+larger scope (bringing increased execution requirements).
+
+[Integration Tests](#integration-testing) verify that the well-tested units do 
+not fail when they interact with components they depend on internally (part of 
+the same project), but will stub/mock external dependencies (from other 
+systems/projects). Compared to unit testing, integration tests tend to be harder 
+to execute, since they rely on an environment similar to production (e.g. they 
+run ion the browser). They also tend to be slower than unit tests making it 
+impossible to exercise all possible scenarios.
+
+While Integration Tests looks for the composition within the system, 
+[System Tests](#system-testing) test the integration with the real external 
+dependencies trying to as close as possible to the production use of a client 
+(in the case of DC/OS, against a real cluster). They are even harder to setup 
+than integration tests because you often have to create custom tooling to 
+operate the whole product. This contributes to them being slower and more often 
+prone to fail occasionally, for example, because of network issues like latency.
+
+## Unit Testing
+
+To ensure that individual units of code (functions/methods) return the expected 
+results with different inputs we write Unit Tests.
+
+Think of a `sum` function. When called as `sum(1)` we may expect a return value 
+of `1`. When called as `sum(1, 2)` we may expect a return value of `3`. And when
+ called with no arguments, we may expect the result to fail with an error.
+
+### Running Unit Tests
+
+Before you run any test, make sure you have 
+[set up your environment](/#dcos-installation-instructions). 
+
+Make sure your packages are up to date by running `npm install`, and then run 
+tests with:
+
+```sh
+npm test
+```
+
+Use `test:watch` if you want the tests to run automatically when a file changes:
+
+```sh
+npm run test:watch
+```
+
+You can even pass parameters to the test engine (in this case jest), when you, 
+for instance, want to run a single spec, for example, `MesosStateUtil`:
+
+```sh
+npm run test -- --watch MesosStateUtil
+```
+
+### Example of a Unit Test
```

Comment:
I have the feeling this might become stale after a while (e.g. we could also write this test with snapshot testing, so as a reader I might think we don't do that at all).

Wouldn't it be simpler to direct the reader directly to one of our test directories?

Reply:
I think we should make an effort to keep this up to date when we introduce new techniques.

For snapshot testing, I think we should add a paragraph. 

I think adding the link to one of the test folders is a great idea (do you recommend any?). I don't think those two are exclusive (we can have both this content and the link)

Reply:
You don't need to add a snapshot paragraph, just wanted to address that I think this paragraph is not ultimatively useful and might go stale ^^

Reply:
If we want to keep the example let's rather have a generic example. e.g. let's take the sum example from above.

Reply:
I will put this improving to a generic example as an improvement.